src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                {
+                    // Consider a deconstruction assignment like the following
+                    // (a, b, c) = (x, y, z);
+
+                    // (x, y, z) are evaluated into temps, then the temps are stored to the targets:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // a = temp1;
+                    // b = temp2;
+                    // c = temp3;
+
+                    // In this code path, we have found a case where a, b, and c are non-side-effecting expressions which refer to unique/non-aliased variables.
+                    // In addition, at least one of the expressions x, y, or z actually needed to be evaluated into temps (versus being a constant value where the temp was elided at an earlier phase).
+                    // As an optimization, ensure that assignments from temps to targets happen in the reverse order of effects:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // c = temp3;
+                    // b = temp2;
+                    // a = temp1;
+
+                    // This makes it more likely that the stack optimizer pass will be able to eliminate the temps and replace them with stack push/pops.
+                    // Note also that no optimization can occur if the expression being converted is not a literal, e.g. `(a, b, c) = M()`. Order of access won't affect our ability to eliminate temps in that case.
+                    effects.assignments.ReverseContents();
+                }
+
+                visitedSymbols.Dispose();
+            }
+
+            static bool canReorderTargetAssignments(ArrayBuilder<Binder.DeconstructionVariable> targets, ref TemporaryArray<Symbol> visitedSymbols)
+            {
+                // If we know all targets refer to distinct variables, then we can reorder the assignments.
+                // We avoid doing this in any cases where aliasing could occur, e.g.:
+                // var y = 1;
+                // ref var x = ref y;
+                // (x, y) = (a, b);
+
+                foreach (var target in targets)
+                {
+                    if (target.Single is { } single)
+                    {
+                        if (single is not (BoundLocal { LocalSymbol.RefKind: RefKind.None }
+                                or BoundParameter { ParameterSymbol.RefKind: RefKind.None }
+                                or BoundDiscardExpression))
+                        {
+                            // This deconstruction assigns to a target which is not sufficiently simple.
+                            // We can't verify that the deconstruction does not use any aliases to variables.
+                            return false;
+                        }
+
+                        var expressionSymbol = single.ExpressionSymbol;
```

Comment:
>ExpressionSymbol [](http://example.com/codeflow?start=54&length=16)

I am very uncomfortable with usage of  ```ExpressionSymbol``` API outside of SemanticModel. It was added to support SemanticModel and is implemented with that in mind. We are enumerating all interesting nodes above and can easily extract relevant symbols from them with a very small refactoring.  Also, there is no point to add/lookup anything for BoundDiscardExpression.

Reply:
Have corrected this. Thanks